### PR TITLE
Adding support for actions

### LINKED ISSUE #9: Support resource actions
When fetching resources (either via a collection search or directly a resource/:id), we populate the actions similar to the collection.  

We need to define the methods for the particular resource dynamically (method_missing) based on those actions so we can support, vm.start, vm.stop  and generically support parameters which are then sent up with the post to the server.

```
POST /api/vms/:id
```

```
{
  "action" : "start",
  "resource" : {
     data
  }
}
```

### LINKED ISSUE #10: Support collection actions
With collection actions populated when collections are searched, we need to define the methods to the collection based on those actions (method_missing), i.e. vms.start, etc.

This would allow us to do the create on the collection:

i.e. 
```
users.create(....)
```

we can also alias new to that.

```
POST /api/users
```

```
{
   data
}
```

Also, this should also allow for the bulk operations.

i.e.

vms.start

```
POST /api/vms
```

```
{
  "action" : "start",
  "resources" : [
     {  "href" : "http://localhost:3000/api/vms/1" },
     {  "href" : "http://localhost:3000/api/vms/2" },
    ...
   ]
}
```

### PR DESCRIPTION
**Pull request by abellotti**
- Resource Actions
    - Collection Actions
    - Auto-fetch available collection actions upon method invokation.
    - Refetch resource if actions undefined upon method invokation.
    - Define hrefs if not yet defined by the server.
    - Added support for ActionResult

Fixes #9 
Fixes #10

### CODE REVIEW

**Comment by abellotti** [comment]
WIP - marked as WIP as it depends on #24

File: lib/manageiq/api/client/collection.rb [hunk 1]
```diff
@@ -117,6 +133,41 @@ def order_parameters_from_query_relation(option)
...
+            res = res["results"].collect { |result| action_result(result) }
+          end
+          res
+        end
+
+        def action_body(action_name, *args, &block)
+          args = args.flatten
+          args = args.first if args.size == 1 && args.first.kind_of?(Hash)
+          args = {} if args.blank?
+          block_data = block ? yield(block) : {}
```
Comment (Fryguy) [inline lib/manageiq/api/client/collection.rb]: This should be `block_data = block ? block.call : {}`

File: lib/manageiq/api/client/mixins/action_mixin.rb [hunk 2]
```diff
@@ -7,15 +7,31 @@ def clear_actions
     @actions = []
   end
 
+  def actions_present?
+    @actions.present? if @actions
```
Comment (Fryguy) [inline lib/manageiq/api/client/mixins/action_mixin.rb]: You don't need the if part

File: lib/manageiq/api/client/mixins/action_mixin.rb [hunk 3]
```diff
@@ -7,15 +7,31 @@ def clear_actions
...
+  end
+
   def fetch_actions(resource_hash)
     @actions = Array(resource_hash["actions"]).collect { |action| ManageIQ::API::Client::Action.new(action) }
   end
 
+  def find_action(action)
+    @actions.detect { |a| a.name == action.to_s } if @actions
```
Comment (Fryguy) [inline lib/manageiq/api/client/mixins/action_mixin.rb]: Minor, but you might want to do the to_s outside of the loop

File: lib/manageiq/api/client/mixins/action_mixin.rb [hunk 4]
```diff
@@ -7,15 +7,31 @@ def clear_actions
...
+  end
+
+  def action_defined?(action)
+    find_action(action) ? true : false
```
Comment (Fryguy) [inline lib/manageiq/api/client/mixins/action_mixin.rb]: Prefer `!!` or just don't even do the convert bool at all

File: lib/manageiq/api/client/resource.rb [hunk 5]
```diff
@@ -27,21 +27,57 @@ def self.subclass(name)
...
+
+        private
+
+        def exec_action(name, args = {}, &block)
+          raise "Action #{name} parameters must be a hash" if !args.kind_of?(Hash)
+          action = find_action(name)
+          res = client.send(action.method, URI(action.href)) do
+            body = { "action" => action.name }
+            resource = args.dup
+            resource.merge!(yield(block)) if block
```
Comment (Fryguy) [inline lib/manageiq/api/client/resource.rb]: Another `yield(block)` -> `block.call`

File: lib/manageiq/api/client/mixins/action_mixin.rb (hunk 4 above)
Comment (abellotti) [inline lib/manageiq/api/client/mixins/action_mixin.rb]: was going to go with !!, but Rubocop was picky :)  so going with second option.

File: lib/manageiq/api/client/resource.rb (hunk 5 above)
Comment (abellotti) [inline lib/manageiq/api/client/resource.rb]: Thanks for catching those ❤️

**Comment by carbonin** [comment]
:tada: Awesome! Great job @abellotti @Fryguy

**Pull request merged**